### ISSUE TITLE
Dial DNS lookup error not retried

### ISSUE THREAD

**Issue by a-canya** [aws/aws-sdk-go-v2#2295]
### Describe the bug

Using the standard retryer (`retry.NewStandard()`), I have observed that the following error is retried:

```
https response error StatusCode: 0, RequestID: , HostID: , request send failed, Put "https://MY_BUCKET.s3.us-east-1.amazonaws.com/MY_FILE?x-id=PutObject": dial tcp 1.2.3.4:443: i/o timeout
```

whereas this error is not retried:

```
https response error StatusCode: 0, RequestID: , HostID: , request send failed, Put "https://MY_BUCKET.s3.us-east-1.amazonaws.com/MY_FILE?x-id=PutObject": dial tcp: lookup MY_BUCKET.s3.us-east-1.amazonaws.com: i/o timeout
```

### Expected Behavior

Both errors are retried.

### Current Behavior

Error with DNS  lookup is not retried.

### Reproduction Steps

I don't know how to reproduce the error. Happened in production a few times.

### Possible Solution

In https://github.com/aws/aws-sdk-go-v2/blob/main/aws/retry/retryable_error.go#L100, there is a cast to *net.DNSError which was introduced in #2083.

```go
	case errors.As(err, &dnsError):
		// NXDOMAIN errors should not be retried
		retryable = !dnsError.IsNotFound && dnsError.IsTemporary
```

This has several issues. `IsTemporary` field is only guaranteed to be valid if true. There is also a method (`Temporary()`) which also considers that the error might be a time out (which I think might be my case). For reference see [net.DNSError](https://pkg.go.dev/net#DNSError). In any case, if false the code should fall through to the other checks.

### AWS Go SDK V2 Module Versions Used

v1.21.0

### Compiler and Version used

go version go1.20.2 linux/amd64

### Operating System and version

alpine:latest

**Comment by a-canya** [comment]
I'm working on [a fix](https://github.com/a-canya/aws-sdk-go-v2/tree/dns-err-not-retried). I will create a PR when I can validate in prod that this fixes our issue.

**Comment by RanVaknin** [comment]
Hi @a-canya ,

Thanks for reaching out. 
While the two errors are in fact DNS related errors, they occur within different areas of the networking stack.

```
https response error StatusCode: 0, RequestID: , HostID: , request send failed, Put "https://MY_BUCKET.s3.us-east-1.amazonaws.com/MY_FILE?x-id=PutObject": dial tcp 1.2.3.4:443: i/o timeout
```
This error indicates that DNS resolution succeeds but not able to establish a connection. This is often for transient reasons. From an SDK design perspective we want to retry connection timeouts since they are often solved with a retry.

```
https response error StatusCode: 0, RequestID: , HostID: , request send failed, Put "https://MY_BUCKET.s3.us-east-1.amazonaws.com/MY_FILE?x-id=PutObject": dial tcp: lookup MY_BUCKET.s3.us-east-1.amazonaws.com: i/o timeout
```
This one is a DNS lookup failure, which often indicates more persistent issues than mere connection timeouts. Hence its not retryable by default.

The SDK tries to accommodate transient error by default, and not general networking errors. It aims to be an isolated layer on top of the standard go library for networking purposes and let customers configure their own networking stack errors (perhaps with a custom retry strategy implemented by the user)

Im inclined to close this issue but if you have any clarification / case for why this should be changed, I'd be happy to discuss it with the team. 

Thanks,
Ran~

**Comment by lucix-aws** [comment]
FYI @a-canya I've taken a look at your patch. I'm fine with taking the change to look at `Temporary()` instead of `IsTemporary` but I'm not sure I understand why the specific "is dns error" path needed to be broken out of the main `switch` expression - they _seem_ behaviorally equivalent to me.

**Comment by lucix-aws** [comment]
OK I see that now. There is definitely a subtle difference between terminating at `retryable = dnsError.Temporary()` and returning on it only _if_ `Temporary()` is true since it being false isn't conclusive. Thanks for clarifying.

Your patch looks solid then if you'd like to PR it - I imagine we'd take it regardless of whether it actually fixes your issue, since you've made a real behavioral improvement regardless.

**Comment by a-canya** [comment]
Hi, thanks @lucix-aws ! I will open the PR.

FYI the error happened again in our system and I managed to get more logs:

```
&net.OpError{Op:"dial", Net:"tcp", Source:net.Addr(nil), Addr:net.Addr(nil), Err:(*net.DNSError)(0xc025604840)}
&net.DNSError{Err:"i/o timeout", Name:"MY_BUCKET.s3.us-east-1.amazonaws.com", Server:"", IsTimeout:true, IsTemporary:false, IsNotFound:false}
```

As suspected, `IsTemporary` is false but `IsTimeout` is true.

**Comment by lucix-aws** [comment]
Resolved by #2300
